src/NodeServices/PythonServices.cs

```diff
@@ -0,0 +1,338 @@
+            {
+                evaluatorClass = CPythonEvaluatorClass;
+                evaluationMethod = CPythonEvaluationMethod;
+                return;
+            }
+
+            // Throwing at the compilation stage is handled as a non-retryable error by the Dynamo engine.
+            // Instead, we want to produce an error at the evaluation stage, so we can eventually recover.
+            // We handle this by providing a dummy Python evaluator that will evaluate to an error message.
+            evaluatorClass = DummyEvaluatorClass;
+            evaluationMethod = DummyEvaluatorMethod;
+        }
+
+        private void AssemblyLoadEventHandler(object sender, AssemblyLoadEventArgs args)
+        {
+            PythonEngineManager.Instance.ScanPythonEngine(args.LoadedAssembly);
+        }
+
+        private void ScanPythonEngine(Assembly assembly)
+        {
+            if (assembly == null)
+            {
+                return;
+            }
+
+            if (!AvailableEngines.Any(x => x.Version == PythonEngineVersion.IronPython2))
+            {
+                var engine = CreateEngineProxy(assembly, PythonEngineVersion.IronPython2);
+                if (engine != null)
+                {
+                    AvailableEngines.Add(engine);
+                }
+            }
+
+            if (!AvailableEngines.Any(x => x.Version == PythonEngineVersion.CPython3))
+            {
+                var engine = CreateEngineProxy(assembly, PythonEngineVersion.CPython3);
+                if (engine != null)
+                {
+                    AvailableEngines.Add(engine);
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// This class is intended to wrap or act as a proxy for many different python engine versions
+    /// </summary>
+    [SupressImportIntoVM]
+    public class PythonEngineProxy : IDisposable
+    {
+        /// <summary>
+        /// The version of the Python Engine connected to this proxy instance
+        /// </summary>
+        public readonly PythonEngineVersion Version;
+
+        private readonly Type EngineType;
+        private EvaluationStartedEventHandler EvaluationStartedCallback;
+        private EvaluationFinishedEventHandler EvaluationFinishedCallback;
+
+        /// <summary>
+        /// Add an event handler before the Python evaluation begins
+        /// </summary>
+        /// <param name="callback"></param>
+        public void OnEvaluationBegin(EvaluationStartedEventHandler callback)
+        {
+            EvaluationStartedCallback += callback;
+            AddEventHandler(EvaluationStartedCallback.GetType(), nameof(HandleEvaluationStarted));
+        }
+
+        /// <summary>
+        /// Add an event handler after the Python evaluation ends
+        /// </summary>
+        /// <param name="callback"></param>
+        public void OnEvaluationEnd(EvaluationFinishedEventHandler callback)
+        {
+            EvaluationFinishedCallback += callback;
```

Comment:
hmm - so if you call this multiple times will the handler be registered over and over to be called multiple times?

Reply:
do these get cleaned up in the dispose call that uses reflection? If so I think it deserves some comments.

Reply:
not used anymore